Csrf filter improvements

ChristopherSchultz opened the pull request:
Please see https://lists.apache.org/thread/47syblyghh3tromyf6bkvl8q14w70f3x for the initial conversation.

I see some potential improvements for the CSRF prevention filter that will be worthwhile.

Specifically:
1. Non-enforcement mode(s?) to help locate problems without breaking an application
2. URL patterns to ignore when adding CSRF tokens (e.g. static resources) to avoid breaking caching

ChristopherSchultz commented:
My initial testing indicates that caching is working as expected with these changes.

ChristopherSchultz commented:
Commit [e2f78ec](https://github.com/apache/tomcat/pull/681/commits/e2f78eca0c7626303e5e50f1f033770b466f1755) adds nonce-check skipping to the URLs that won't get nonces added to them.

This is necessary if the `<url-pattern>` for the `<filter-mapping>` is set to `/*` as recommended in the documentation and you want to be able to load static resources from your own application :)

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -53,6 +58,25 @@ public class CsrfPreventionFilter extends CsrfPreventionFilterBase {
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".js"));
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".gif"));
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".png"));
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".jpg"));
```
`jpeg`, `jpg` is a DOS problem.

ChristopherSchultz replied:
I don't follow. How is allowing requests for .jpg and .jpeg (not mentioned here) a DOS problem? Feel free to message me privately if something should be kept quiet.

Oh... did you mean Disk Operating System or Denial of Service?

michael-o replied:
The original extension is `.jpeg`. `.jpg` only exists because IBM's Disk Operating System as too stupid to handle more than three chars for file extensions.

ChristopherSchultz replied:
I've used UNIX-like systems for decades and I rarely see `.jpeg`, but I'm happy to add it. I was trying to keep the out-of-the-box processing to a minimum.

ChristopherSchultz replied:
I've added `.jpeg`

isapir replied:
"I never type `.jpeg` because I can get the same money for `.jpg`" - Mark Twain

But sure, it doesn't hurt to add it as others do like the longhand (is that a word? opposite of shorthand) form

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -53,6 +58,25 @@ public class CsrfPreventionFilter extends CsrfPreventionFilterBase {
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".jpg"));
+        defaultNoNonceURLPatterns.add(new SuffixPredicate(".svg"));
+
+        DEFAULT_NO_NONCE_URL_PATTERNS = Collections.unmodifiableList(defaultNoNonceURLPatterns);
```
What about `.ico`?

ChristopherSchultz replied:
Sure. I'm happy to add some obvious things. I thought `.ico` was dead, actually.

markt-asf replied:
Might this be better handled by looking at content type rather than file extension?

michael-o replied:
Is this really dead https://en.wikipedia.org/wiki/Favicon#How_to_use?

ChristopherSchultz replied:
I just assumed everyone used file formats other than `.ico` for their "Favicon". It's not "favicon" that's dead... it's `.ico`. But I can add it to the default. Remember: these are just the _defaults_. Anyone can change them vi configuration.

ChristopherSchultz replied:
> Might this be better handled by looking at content type rather than file extension?

All we have to go on is the path of the file, so I don't think we can check the content-type. Remember, this is primarily being used inside of `CsrfResponseWrapper.encodeURL(String)` where we know nothing about the file other than the path.

ChristopherSchultz replied:
I've added `.ico` but I want to ensure I understand @markt-asf's comment about the content-type before I "resolve" this review.

markt-asf replied:
Use `ServletContext.getMimeType(String)`? My thinking was a filter for `image/*` was a lot more compact and somewhat future proof. You could also exclude `*font*` and so on.

michael-o replied:
Is is possible that an image is created on the fly containing some kind of secret (e.g., QR code, OTP) which should be protected?

ChristopherSchultz replied:
@markt-asf Sure, this could work. I could use a new type of matcher such as `mime:[whatever]` and then use the match based upon MIME type instead of filename. Then you can mix-and-match.

ChristopherSchultz replied:
@michael-o All of this stuff can be configured-around by the user. We are just talking about the defaults, here. If you want to serve static content from `/static/*.jpg` but dynamic files from `/dynamic/*.jpg` which need protection, then you can set up a regular-expression-based check.

An out-of-the-box default should work for "a great many environments" not "every single conceivable environment".

ChristopherSchultz replied:
I think this has been resolved in ef54a1e44fe32aa0ec0fd0559726b214275046aa and d0433b10c98b90cdca573a14c4eb64dff8bdc980.

michael-o replied:
Agree

ChristopherSchultz commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -53,6 +58,25 @@ public class CsrfPreventionFilter extends CsrfPreventionFilterBase {
+
+    private Collection<Predicate<String>> noNoncePatterns = DEFAULT_NO_NONCE_URL_PATTERNS;
+
+    private static final Collection<Predicate<String>> DEFAULT_NO_NONCE_URL_PATTERNS;
```
Sure, I can re-arrange these.

kkolinko commented:
1. There are case-insensitive file systems out there... I wonder whether those default extensions should be treated case-insensitively. (If one is serving a web site from an USB stick or a memory card formatted with FAT? From a CD Drive? It is possible, but rare nowadays.)
2. Add "*.mjs" to the list (see https://bz.apache.org/bugzilla/show_bug.cgi?id=68378 )
3. Documentation: The value in "The default is ..." does not match the actual value of DEFAULT_NO_NONCE_URL_PATTERNS;
4.  Documentation: "Complete regular expression ... Note that patterns cannot contain a comma"
I think if the value starts and ends with a '/'. it would be better to treat it whole as a single RegExp. Commas are useful in RegExes and disallowing them in this case does not look like a benefit.
5. protected boolean skipNonceCheck(HttpServletRequest request) {
It is hard-coded to look for GET. How about a HEAD request?
6. protected boolean skipNonceCheck(HttpServletRequest request) {
Further in that method. "if (!entryPoints.contains(requestedPath)) { return false; }" - note that unless it is an entry point, processing will end here and subsequent lines will not run. I think it was intended to be the opposite.
7. private boolean shouldAddNonce(String url) { ... }
I think that it would make sense to skip adding nonces to the entryPoints.
(As a use case: the front page of Manager web application).
8. It would be good to have some test cases.

ChristopherSchultz commented:
> 1. There are case-insensitive file systems out there... I wonder whether those default extensions should be treated case-insensitively. (If one is serving a web site from an USB stick or a memory card formatted with FAT? From a CD Drive? It is possible, but rare nowadays.)

Fair. Again, I was thinking of trying to minimize the amount of processing required by default.

>     2. Add "*.mjs" to the list (see https://bz.apache.org/bugzilla/show_bug.cgi?id=68378 )

Fair.

>     3. Documentation: The value in "The default is ..." does not match the actual value of DEFAULT_NO_NONCE_URL_PATTERNS;

I will correct this.

>     4. Documentation: "Complete regular expression ... Note that patterns cannot contain a comma"
>        I think if the value starts and ends with a '/'. it would be better to treat it whole as a single RegExp. Commas are useful in RegExes and disallowing them in this case does not look like a benefit.

I suppose I could write a more fully-featured parser, but right now I'm using `String.split(",")` to separate the patterns from each other. If we want to parse `/anything/` including commas, we'll need to be able to recognize `/` within `/.../`, escapes, etc.

I think I might like to save that for a separate PR since this one is complicated enough. WDYT?

>     5. protected boolean skipNonceCheck(HttpServletRequest request) {
>        It is hard-coded to look for GET. How about a HEAD request?

This check pre-dates this PR. I think it should be addressed separately.

>     6. protected boolean skipNonceCheck(HttpServletRequest request) {
>        Further in that method. "if (!entryPoints.contains(requestedPath)) { return false; }" - note that unless it is an entry point, processing will end here and subsequent lines will not run. I think it was intended to be the opposite.

I will review.

>     7. private boolean shouldAddNonce(String url) { ... }
>        I think that it would make sense to skip adding nonces to the entryPoints.
>        (As a use case: the front page of Manager web application).

I think it does not matter much.

>     8. It would be good to have some test cases.

Okay. Would you prefer very targeted unit tests against e.g. the predicates and calls to `HttpServletResponse.encodeURL` or something that includes the whole HTTP request/response, page-generation, etc.?

kkolinko commented:
Re 4: I think that if one is wise enough to write a RegExp, they could use "|" to combine several patterns, and do not really need splitting by comma. Or do you envision a use case, where different types of patterns are used together, and one of them is a regular expression?

I mean: do a .startsWith("/") && .endsWith("/") test before calling String.split(). Skip splitting.

```diff
- if (null == patterns || 0 == patterns.trim().length()) {
+ if (null == patterns || 0 == (patterns = patterns.trim()).length()) {
...
- String values[] = patterns.split(",");
+ String values[] = patterns.startsWith("/") && patterns.endsWith("/") ? new String[]{ patterns } : patterns.split(",");
```

kkolinko commented:
Re 8: Whatever is easier.
(Maybe it will be easier to extract some logic into an utility class and test that utility class. My concern is just that the logic is not trivial, is complicated by nuances like case-insensitivity, and is not tested).

(Existing TestCsrfPreventionFilter class has several "simple" tests. TestRestCsrfPreventionFilter has more substantial tests, using mocks).

ChristopherSchultz commented:
> Re 4: I think that if one is wise enough to write a RegExp, they could use "|" to combine several patterns, and do not really need splitting by comma. Or do you envision a use case, where different types of patterns are used together, and one of them is a regular expression?
> 
> I mean: do a .startsWith("/") && .endsWith("/") test before calling String.split(). Skip splitting.
> 
> ```diff
> - if (null == patterns || 0 == patterns.trim().length()) {
> + if (null == patterns || 0 == (patterns = patterns.trim()).length()) {
> ...
> - String values[] = patterns.split(",");
> + String values[] = patterns.startsWith("/") && patterns.endsWith("/") ? new String[]{ patterns } : patterns.split(",");
> ```

Yes, I was thinking that someone could specify a series of checks like `*.css, /.*includes.*/, *.png`. I suppose if you are going to use a regular expression, maybe the entire pattern should be used since regex is very expensive already.

ChristopherSchultz commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -198,15 +416,27 @@ protected boolean skipNonceCheck(HttpServletRequest request) {
+            for (Predicate<String> p : noNoncePredicates) {
+                if (p.test(requestedPath)) {
+                    if (log.isTraceEnabled()) {
+                        log.trace("Skipping CSRF nonce-check for GET request to no-nonce path " + requestedPath);
```
> No `messages.properties`?

This class does not currently use localized exception messages. I'm happy to do that work in a separate PR. I'm trying not to re-write every line of the source file, and I'm trying to keep things consistent.

michael-o replied:
Agreed.

ChristopherSchultz commented on webapps/docs/config/filter.xml:
```diff
@@ -319,6 +326,34 @@
+        to them. You may not want to add nonces to certain URLs to avoid
+        creating unique URLs which may defeat resource caching, etc.</p>
+
+        <p>There are 3 types of patterns supported:</p>
```
3 == three

Though there are actually 4 (four?), now, so that needs a fix.

Is there a documented preference for spelled-out numerals in Tomcat documentation?

michael-o replied:
Most style guides will likely tell your to which number should be written up. APA, Chicago, etc.

ChristopherSchultz replied:
2ab2317d581fb6657fe02529d5ad55af00161726

ChristopherSchultz commented on webapps/docs/config/filter.xml:
```diff
@@ -291,6 +291,13 @@
+        disabled, the CsrfPreventionFilter will <i>allow all requests</i> and
+        log CSRF failures as DEBUG messages. The default is <b>true</b>,
+        enabling the enforcement of CSRF protections.</p>
+      </attribute>
```
Removing the filter from `web.xml` will not produce log messages for CSRF failures, nor will it add CSRF tokens to URLs produced by the application.

Running in an non-enforcement mode is helpful to collect real-world information about your application without breaking it.

Please see https://lists.apache.org/thread/47syblyghh3tromyf6bkvl8q14w70f3x for the initial conversation, where I make the case for a non-enforcement mode.

michael-o replied:
I see, agreed.

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     * Sets the flag to enforce CSRF protection or just log failures as DEBUG
+     * messages.
+     *
+     * @param enforce <code>true</code> to enforce CSRF protections or
```
Singular or plural? Above is singular

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     * Gets the flag to enforce CSRF protection or just log failures as DEBUG
+     * messages.
+     *
+     * @return <code>true</code> if CSRF protections will be enforced or
```
Same here

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     *         <code>false</code> if all requests will be allowed and
+     *         failures will be logged as DEBUG messages.
+     */
+    public boolean getEnforce() {
```
Is is common throughout Tomcat source code istead of `isEnforce()` for boolean return value?

ChristopherSchultz replied:
There are 308 classes in the Tomcat 11.0.x source tree which contain `boolean isFoo()` and only 141 which contain `boolean getFoo()`. I didn't bother checking is any contain both.

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     * @return A collection of predicates representing the URL patterns.
+     */
+    protected static Collection<Predicate<String>> createNoNoncePredicates(ServletContext context, String patterns) {
+        if (null == patterns || 0 == patterns.trim().length()) {
```
why is `#isEmpty()`?

isapir replied:
In Java 11+ you can use `patterns.isBlank()` which will return true if the string is empty or contains only whitespace

michael-o replied:
Can't be backported

ChristopherSchultz replied:
@michael-o None of this can be back-ported farther than 10.1.x without significant changes. The whole `jakarta.*` namespace needs to change, `Predicate` needs to be defined locally, etc. I'm okay using `String.isBlank` here if anyone is passionate about it. I generally prefer things to be as close as possible across the branches, so I would err on the side of using `null`-check-plus-zero-length-check until we dump Tomcat 8.5.x.

ChristopherSchultz replied:
> why is `#isEmpty()`?

I'm not sure I understand this, @michael-o

isapir replied:
Passionate is a strong word, so I can't say that I'm "passionate" about using String.isBlank(), but OTOH I do like seeing modern Java in the source code where applicable, so I'm a +1 on that one.

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+        public boolean test(String t) {
+            String mimeType = context.getMimeType(t);
+
+            return predicate.test(mimeType);
```
Does predicate handle `null`?

ChristopherSchultz replied:
Is it often that the application will invoke `HttpServletResponse.encodeURL(null)`?

michael-o commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+
+        @Override
+        public boolean test(String t) {
+            return t.endsWith(this.prefix);
```
`startsWith()`?

isapir replied:
Good catch @michael-o :+1:

isapir commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
+     */
+    protected static Collection<Predicate<String>> createNoNoncePredicates(ServletContext context, String patterns) {
+        if (null == patterns || 0 == patterns.trim().length()) {
+            return null;
```
What do you think about returning `Collections.emptyList()` instead of null? Then you can use it in iterators with no issues and avoid having to check for nulls / NPEs

ChristopherSchultz replied:
I think this is a matter of taste these days. I usually prefer `null` to empty collections just because the null-check is far faster than creating an iterator from an empty collection, then iterating zero times over it. All kinds of control-flow is skipped with a simple null-check.

It does make the code a little cleaner, though. Does anyone else want to weigh-in?

isapir commented on java/org/apache/catalina/filters/CsrfPreventionFilter.java:
```diff
@@ -110,45 +285,70 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
-                }
+                    if (enforce(req, requestedPath)) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Rejecting request for " + getRequestedPath(req) + ", session " +
```
Looks like quite a few repetitions of log.debug blocks with very similar content.  Might be a good opportunity for a private function?

ChristopherSchultz replied:
If you really think it needs it. I don't find the control-flow difficult to follow, here, with the log messages in there. I think further obfuscating the log messages will just generate more code.

isapir replied:
I personally think that the code is cleaner and easier to maintain when repetitive blocks are encapsulated in a function, but do I "really think it needs it"? Nahh - your code, your decision :)

ChristopherSchultz commented:
> > Re 4: I think that if one is wise enough to write a RegExp, they could use "|" to combine several patterns, and do not really need splitting by comma.
>
> I suppose if you are going to use a regular expression, maybe the entire pattern should be used since regex is very expensive already.

After thinking about this for a while, I agree with @kkolinko on this one: a regular expression is going to be expensive enough to run that the user may as well implement the whole matching pattern in a single regular expression instead of having Tomcat chop it up into little pieces and potentially run multiple regular expression matches for every URL-check.

I will change the implementation to allow *either* a single regular expression (offset by leading `/` and trailing `/` characters) *or* a series of simpler comma-separated matchers.

michael-o reviewed (approved):
I have no objections and trust you here.
